src/test/java/org/junit/tests/assertion/AssertionTest.java

```diff
@@ -717,6 +717,27 @@ public void stringNullAndNullWorksToo() {
         assertEquals("null", null);
     }
 
+    private static class NullToString {
+        @Override
+        public String toString() {
+            return null;
+        }
+    }
+
+    @Test
+    public void nullToString() {
+        try {
+            assertEquals(new NullToString(), new NullToString());
+        } catch (AssertionError e) {
+            assertEquals("expected: org.junit.tests.assertion.AssertionTest$NullToString<null> but "
+                            + "was: org.junit.tests.assertion.AssertionTest$NullToString<null>",
+                    e.getMessage());
+            return;
+        }
+
+        fail("Failed on assertion.");
```

Comment:
Calling `fail` from inside of a `try` block that has `catch (AssertionError` doesn't do what many people think it does, so I'm strongly against that option. Maybe I'll find time to fix the tests that do that.

I also dislike using  `@Test(expected = Something.class)` for any test method where there is more than one call in the body, since it isn't clear which method should throw the exception. Because using `expected` should usually only be used for single-method tests, I tend to completely avoid it in my personal projects, but it's usage is very common in the JUnit tests.

Throwing a custom exception inside the `try` block for something that should be an assertion error is a bit odd, so I'm not thrilled with that option.

I personally strongly in favor of the `return` and `fail` idiom for any tests that verify that a test method throws `AssertionError` unless the test can be written with one line of code (in which case  `@Test(expected = AssertionError.class)` is fine).

Luckily, with Java 8 we can all use lambdas so this will all be moot some day.